### PR TITLE
Added JSON serialization and multi-recipient support to JWE

### PR DESCRIPTION
**Pull request by spivachuk**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
    - Added JSON serialization and multi-recipient support to JWE
- [ ] Code style update
- [ ] Refactor
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

---

- [x] You consent that the copyright of your pull request source code belongs to Authlib's author.

### CODE REVIEW

File: authlib/jose/drafts/_jwe_algorithms.py [hunk 1]
```diff
@@ -32,6 +32,15 @@ def prepare_key(self, raw_data):
             return raw_data
         return ECKey.import_key(raw_data)
 
+    def generate_preset(self, enc_alg, key):
+        epk = key.generate_key(key['crv'], is_private=True)
```
Comment (ashcherbakov) [inline authlib/jose/drafts/_jwe_algorithms.py]: Should we call `self._generate_ephemeral_key(key)` instead?

File: authlib/jose/drafts/_jwe_algorithms.py [hunk 2]
```diff
@@ -88,19 +97,27 @@ def deliver_at_recipient(self, recipient_key, sender_static_pubkey, sender_ephem
...
+    def generate_keys_and_prepare_headers(self, enc_alg, key, sender_key, preset=None):
         if not isinstance(enc_alg, CBCHS2EncAlgorithm):
             raise InvalidEncryptionAlgorithmForECDH1PUWithKeyWrappingError()
 
-        epk = self._generate_ephemeral_key(key)
-        cek = enc_alg.generate_cek()
-        h = self._prepare_headers(sender_key, epk)
+        if preset and 'epk' in preset:
+            epk = preset['epk']
+            h = {}
```
Comment (ashcherbakov) [inline authlib/jose/drafts/_jwe_algorithms.py:112]: Should we get the header from `preset`?

File: authlib/jose/rfc7516/jwe.py [hunk 3]
```diff
@@ -139,6 +141,145 @@ def serialize_compact(self, protected, payload, key, sender_key=None):
...
+
+        if not keys:
+            raise ValueError("No keys have been provided")
+
+        shared_header = JWESharedHeader.from_dict(header_obj)
+
+        recipients = header_obj.get('recipients')
+        if recipients is None:
+            recipients = [{} for _ in keys]
+        for i in range(len(recipients)):
```
Comment (ashcherbakov) [inline authlib/jose/rfc7516/jwe.py:266]:
```suggestion
        for recipient in recipients:
```

File: authlib/jose/rfc7516/jwe.py [hunk 4]
```diff
@@ -139,6 +141,145 @@ def serialize_compact(self, protected, payload, key, sender_key=None):
...
+        if len(keys) != len(recipients):
+            raise ValueError("Count of recipient keys {} does not equal to count of recipients {}"
+                             .format(len(keys), len(recipients)))
+
+        # step 1: Prepare algorithms & key
+        alg = self.get_header_alg(shared_header)
+        enc = self.get_header_enc(shared_header)
+        zip_alg = self.get_header_zip(shared_header)
+
+        self._validate_sender_key(sender_key, alg)
```
Comment (ashcherbakov) [inline authlib/jose/rfc7516/jwe.py:283]: Do we validate recipient keys and check that sender and recipient keys are on the same curve?

File: authlib/jose/rfc7516/jwe.py [hunk 5]
```diff
@@ -197,6 +338,110 @@ def deserialize_compact(self, s, key, decode=None, sender_key=None):
             payload = decode(payload)
         return {'header': protected, 'payload': payload}
 
+    def deserialize_json(self, obj, key, decode=None, sender_key=None):
```
Comment (ashcherbakov) [inline authlib/jose/rfc7516/jwe.py:480]: In many cases we will need to parse the message first, get the recipient key IDs, and only then call descrypt/deserialize for the found recipient key values. So, we will need a separate method to parse the message.

File: authlib/jose/rfc7516/jwe.py [hunk 6]
```diff
@@ -197,6 +338,110 @@ def deserialize_compact(self, s, key, decode=None, sender_key=None):
...
+
+        tag = extract_segment(to_bytes(obj['tag']), DecodeError, 'authentication tag')
+
+        shared_header = JWESharedHeader(protected, unprotected)
+
+        alg = self.get_header_alg(shared_header)
+        enc = self.get_header_enc(shared_header)
+        zip_alg = self.get_header_zip(shared_header)
+
+        self._validate_sender_key(sender_key, alg)
```
Comment (ashcherbakov) [inline authlib/jose/rfc7516/jwe.py:527]: Do we validate recipient keys and check that sender and recipient keys are on the same curve?

File: authlib/jose/rfc7516/jwe.py [hunk 7]
```diff
@@ -197,6 +338,110 @@ def deserialize_compact(self, s, key, decode=None, sender_key=None):
...
+            return alg.unwrap(enc, ek, header, key, sender_key, tag)
+
+        def _unwrap_with_sender_key_and_without_tag(ek, header):
+            return alg.unwrap(enc, ek, header, key, sender_key)
+
+        def _unwrap_without_sender_key_and_tag(ek, header):
+            return alg.unwrap(enc, ek, header, key)
+
+        def _unwrap_for_some_recipient(unwrap_func):
+            for recipient in recipients:
```
Comment (ashcherbakov) [inline authlib/jose/rfc7516/jwe.py:565]: I believe we must be able to provide a key with a key ID, so that we can avoid that iteration, and try to decrypt for the given key only.

File: authlib/jose/rfc7518/jwe_algs.py [hunk 8]
```diff
@@ -187,6 +213,15 @@ def prepare_key(self, raw_data):
             return raw_data
         return ECKey.import_key(raw_data)
 
+    def generate_preset(self, enc_alg, key):
+        epk = key.generate_key(key['crv'], is_private=True)
```
Comment (ashcherbakov) [inline authlib/jose/rfc7518/jwe_algs.py]: Should we call self._generate_ephemeral_key(key) instead?

File: tests/core/test_jose/test_jwe.py [hunk 9]
```diff
@@ -886,6 +886,73 @@ def test_ecdh_1pu_jwe_with_okp_keys_in_key_agreement_with_key_wrapping_mode(self
                 rv = jwe.deserialize_compact(data, bob_key, sender_key=alice_key)
                 self.assertEqual(rv['payload'], b'hello')
 
+    def test_ecdh_1pu_jwe_json_serialization(self):
```
Comment (ashcherbakov) [inline tests/core/test_jose/test_jwe.py]: Please add more tests for JSON serialization and multi-recipient.
Please add tests for the test vectors from ECDH-1PU and ECDH-ES where JSON serialization is used.

Comment (ashcherbakov) [reply tests/core/test_jose/test_jwe.py]: Also, please add tests with edge cases, where unexpected parameters are passed:
- keys from different curves
- invalid/unexpected/missing values in the input header
- etc.

File: authlib/jose/drafts/_jwe_algorithms.py (hunk 2 above)
Comment (spivachuk) [reply authlib/jose/drafts/_jwe_algorithms.py:112]: No, because here we construct specific recipient's unprotected header.

File: authlib/jose/rfc7516/jwe.py (hunk 3 above)
Comment (spivachuk) [reply authlib/jose/rfc7516/jwe.py:266]: No, otherwise replacement of a recipient being None with an empty dict will take no effect.

File: authlib/jose/rfc7516/jwe.py (hunk 7 above)
Comment (spivachuk) [reply authlib/jose/rfc7516/jwe.py:565]: Added such logic.

File: authlib/jose/drafts/_jwe_algorithms.py (hunk 1 above)
Comment (spivachuk) [reply authlib/jose/drafts/_jwe_algorithms.py]: Fixed.

File: authlib/jose/rfc7518/jwe_algs.py (hunk 8 above)
Comment (spivachuk) [reply authlib/jose/rfc7518/jwe_algs.py]: Fixed.

File: authlib/jose/rfc7516/jwe.py (hunk 5 above)
Comment (spivachuk) [reply authlib/jose/rfc7516/jwe.py:480]: Added `parse_json` method.

File: tests/core/test_jose/test_jwe.py (hunk 9 above)
Comment (spivachuk) [reply tests/core/test_jose/test_jwe.py]: Done.

File: authlib/jose/rfc7516/jwe.py (hunk 6 above)
Comment (spivachuk) [reply authlib/jose/rfc7516/jwe.py:527]: This is done by the underlying crypto library.

File: authlib/jose/rfc7516/jwe.py (hunk 4 above)
Comment (spivachuk) [reply authlib/jose/rfc7516/jwe.py:283]: This is done by the underlying crypto library.

File: tests/core/test_jose/test_jwe.py [hunk 10]
```diff
@@ -1139,3 +2197,384 @@ def test_dir_alg_c20p(self):
...
+        protected = {'alg': 'ECDH-1PU+A128KW', 'enc': 'A128CBC-HS256'}
+        header_obj = {'protected': protected}
+
+        data = jwe.serialize(header_obj, b'hello', bob_key, sender_key=alice_key)
+        self.assertIsInstance(data, dict)
+
+        data_as_string = json.dumps(data)
+
+        rv = jwe.deserialize(data_as_string, bob_key, sender_key=alice_key)
+        self.assertEqual(rv['payload'], b'hello')
```
Comment (ashcherbakov) [inline tests/core/test_jose/test_jwe.py:3076]: 2580 lines...
Did you consider split the test into multiple files?

File: tests/core/test_jose/test_jwe.py [hunk 11]
```diff
@@ -405,6 +420,229 @@ def test_ecdh_es_jwe_with_okp_key_in_key_agreement_with_key_wrapping_mode(self):
...
+        self.assertEqual(rv_at_charlie['header']['aad'], jwe_aad)
+        self.assertEqual(rv_at_charlie['payload'], payload)
+
+    def test_ecdh_es_jwe_with_json_serialization_when_kid_is_specified(self):
+        jwe = JsonWebEncryption()
+
+        bob_key = OKPKey.import_key({
+            "kty": "OKP",
+            "crv": "X25519",
+            "kid": "bob-key-2",
```
Comment (ashcherbakov) [inline tests/core/test_jose/test_jwe.py:791]: Is the only way to specify a KID for decryption is within a JWK object? If so, it may not work for DIDComm.
See https://www.w3.org/TR/did-spec-registries/#jsonwebkey2020
There JWK `kid` is a relative value, we can not resolve a DID based on it. What we will specify as a `kid` in per -recipient header is verification method's `id`, that is DID URL for which we can resolve the DID and find the corresponding key.

File: tests/core/test_jose/test_jwe.py [hunk 12]
```diff
@@ -1139,3 +2197,384 @@ def test_dir_alg_c20p(self):
...
+
+        protected = {'alg': 'ECDH-1PU+A128KW', 'enc': 'A128CBC-HS256'}
+        header_obj = {'protected': protected}
+
+        data = jwe.serialize(header_obj, b'hello', bob_key, sender_key=alice_key)
+        self.assertIsInstance(data, dict)
+
+        rv = jwe.deserialize(data, bob_key, sender_key=alice_key)
+        self.assertEqual(rv['payload'], b'hello')
+
```
Comment (ashcherbakov) [inline tests/core/test_jose/test_jwe.py:3060]: Please add the following tests:
1) Decrypt test: Some random values in JSON for decryption (not from the spec). Expected: ignored and decrypt successful?
2) Decrypt test: ciphertext is invalid. Expected: exception
3) Decrypt test: no epk in protected header.  Expected: exception
4) Decrypt test: `encrypted_key` value for the the given recipient key is invalid.  Expected: exception
5) Decrypt test: `encrypted_key` value for the the given recipient key is valid, but for another recipient not passed to decrypt method it is not valid  Expected: success

File: tests/core/test_jose/test_jwe.py (hunk 11 above)
Comment (spivachuk) [reply tests/core/test_jose/test_jwe.py:791]: Supported a separately provided recipient kid in deserialize methods of `JsonWebEncryption` class.

File: tests/core/test_jose/test_jwe.py (hunk 12 above)
Comment (spivachuk) [reply tests/core/test_jose/test_jwe.py:3060]: Done except for Item 3 because modifying protected header of JWE message (removing `epk` header field before the message decryption) breaks the message consistency (i.e. consistency with AAD, ciphertext, tag).

File: tests/core/test_jose/test_jwe.py [hunk 13]
```diff
@@ -238,6 +239,205 @@ def test_aes_gcm_jwe_invalid_key(self):
...
         )
 
+    def test_serialize_compact_fails_if_header_contains_unknown_field_while_private_fields_restricted(self):
+        jwe = JsonWebEncryption(private_headers=[])
+        key = OKPKey.generate_key('X25519', is_private=True)
+
+        protected = {
+            "alg": "ECDH-ES+A128KW",
+            "enc": "A128GCM",
+            "foo": "bar"
```
Comment (ashcherbakov) [inline tests/core/test_jose/test_jwe.py:250]: Do we really need to fail here? Does JOSE spec explicitly restrict any values besides the ones listed there?
BTW In Numbus unknown fields are just ignored, there is no failure.

File: tests/core/test_jose/test_jwe.py [hunk 14]
```diff
@@ -238,6 +239,205 @@ def test_aes_gcm_jwe_invalid_key(self):
...
+        }
+
+        self.assertRaises(
+            InvalidHeaderParameterNameError,
+            jwe.serialize_compact,
+            protected, b'hello', key
+        )
+
+    def test_serialize_json_fails_if_protected_header_contains_unknown_field_while_private_fields_restricted(self):
+        jwe = JsonWebEncryption(private_headers=[])
```
Comment (ashcherbakov) [inline tests/core/test_jose/test_jwe.py]: What does `private_fields_restricted` means here?

Comment (spivachuk) [reply tests/core/test_jose/test_jwe.py]: The mode of `JsonWebEncryption` class which restricts the set of allowed private header fields. Here the set of allowed private header fields is empty (since `private_headers` argument of `JsonWebEncryption` constructor is empty).

File: tests/core/test_jose/test_jwe.py (hunk 13 above)
Comment (spivachuk) [reply tests/core/test_jose/test_jwe.py:250]: This behavior is controlled by `private_headers` parameter of `JsonWebEncryption` constructor.

**Review by ashcherbakov — APPROVED** [review]
